### PR TITLE
subsys: power: initialize states as disabled

### LINKED ISSUE #11908: Power Manager does not handle K_FOREVER properly
**Describe the bug**
Power Manger will go into deep sleep if `sys_pm_policy_next_state` is called with K_FOREVER even if there is still work to be done.

**To Reproduce**
1. Checkout v1.13
2. Build `/samples/subsys/power` with ` cmake -DBOARD=nrf52_pca20020 -GNinja ../..` (defaults)
3. run it on a thingy 52

**Expected behavior**
See App output on console:
```
***** Booting Zephyr OS zephyr-v1.13.0 ****

***OS Power Management Demo on arm****
Demo Description
Application creates Idleness, Due to which System Idle Thread is
scheduled and it enters into various Low Power States.
```

**Actual behavior**
Only `Entering Low Power state (3)` is printed. device is in deep sleep and no possibility to wake it up. `main` is never reached

**Impact**
Big we have to implement own Power Manager / Patch policy.c which is scary

**Screenshots or console output**
see above

**Environment (please complete the following information):**
 - OS: Debian Linux
 - Toolchain Zephyr SDK
 - `<v1.13.0>`

**Workaround**
The problem seems to be with the handling of `K_FOREVER` following diff will mitigate problem but probably has other issues:
```
diff --git a/subsys/power/policy.c b/subsys/power/policy.c
index ef204e1dde..5eb32965ec 100644
--- a/subsys/power/policy.c
+++ b/subsys/power/policy.c
@@ -72,7 +72,7 @@ int sys_pm_policy_next_state(s32_t ticks, enum power_states *pm_state)
 {
        int i;
 
-       if ((ticks != K_FOREVER) && (ticks < pm_policy[0].min_residency)) {
+       if ((ticks == K_FOREVER) || (ticks < pm_policy[0].min_residency)) {
                LOG_ERR("Not enough time for PM operations: %d\n", ticks);
                return SYS_PM_NOT_HANDLED;
        }

```

### LINKED ISSUE #20775: sys_pm_ctrl: fragility in managing state control
subsys/power supports the option (via `CONFIG_SYS_PM_STATE_LOCK`) to enable and disable  specific power states.  State is managed with an array of `atomic_t` instances.

A request to disable a state corresponds to an atomic increment.

A request to enable a state corresponds to an atomic decrement.

The state is considered enabled when the counter is zero, which is the initial condition.

This approach results in some unusual behavior:

* `sys_pm_ctrl_enable_state(S)` when S is enabled decrements the counter to -1 thereby disabling the state.

* `sys_pm_ctrl_disable_state(S)` when S is disabled means that a subsequent `sys_pm_ctrl_enable_state(S)` may not enable the state.

The code has asserts to detect for underflow and overflow but since assertions are not generally enabled "misuse" is unlikely to be found.

It may be that this behavior is intended.  If so it needs to be documented.

### PR DESCRIPTION
**Pull request by mnkp**
This commit makes behavior enabled previously via SYS_PM_STATE_LOCK Kconfig option the default and removes the option. Upon boot all power states are disabled and need to be explicitely enabled by the application or the power management subsystem by calling sys_pm_ctrl_enable_state function.

More information is provided in the updated documentation of `sys_pm_ctrl_enable_state()`, `sys_pm_ctrl_disable_state()` functions.

All in-tree references to SYS_PM_STATE_LOCK were removed.

Fixes #11908, 
Fixes #20775 
Partially #20792.

### CODE REVIEW

**Comment by jhedberg** [comment]
@mnkp any plans to continue work on this PR? At least one of the issue it's marked as (potentially) fixing is of medium priority, hence it's of interest for Zephyr 2.2.

File: subsys/power/pm_ctrl.c [hunk 1]
```diff
@@ -14,15 +14,15 @@
 #include <logging/log.h>
 LOG_MODULE_DECLARE(power);
 
-static atomic_t power_state_disable_count[SYS_POWER_STATE_MAX];
+static atomic_t power_state_enable_count[SYS_POWER_STATE_MAX];
```
Comment (mnkp) [inline subsys/power/pm_ctrl.c]: The functionality of the reworked code is meant to be identical with the only difference that the power states are disabled by default. Previous code was actually easier to understand, however would require adding `SYS_INIT` macro to call a routine that would initialize `power_state_disable_count` array with ones. Also an option.

**Comment by mnkp** [comment]
@wentongwu could you have a look at the PR.

**Review by wentongwu — COMMENTED** [review]
yeah, this PR can prevent system off unexpectedly from being entered on long timeout or sleep forever. But it need current working platforms to do some extra initialization. 
Can we separate system off state from deep sleep states as a individual and specific state, and add SYS_INIT currently in pm_policy.c to do the initialization which will disable system off state if it has been enabled.
Further I think we need define the pm policy to a structure to help more abstract and flexible.

**Review by pabigot — DISMISSED** [review]
In the as-is situation states are enabled by default, represented by a zero value of an atomic counter.  Any number of clients can request a disable, which (increase the count), and having done so can later re-enable (decrease the count).  So as long as at least one disable is registered the state remains disabled.

In the proposed situation states are disabled by default, represented by a zero value of an atomic counter.  During power management initialization the state is expected to be enabled (increasing the count) where appropriate.  But only one client can request it be enabled ("Any power state can be enabled only once."), which is enforced by an assert that the count never exceeds 1.

This seems to require cross-client coordination: if three drivers have a dependency on a system only one can enable it.  The documentation also specifies that a client can only enable if has previously disabled, which can't be true because it starts disabled.  Is each driver independently supposed to either enable or disable all power-managed elements during startup?  If so then the assert `v < 1` in `sys_pm_ctrl_enable_state()` isn't right.

So I'm not sure how this is supposed to be used.  At a minimum the inconsistencies in the documentation should be updated.

**Review by mnkp — COMMENTED** [review]
This PR doesn't change the usage model of `sys_pm_ctrl_disable_state` / `sys_pm_ctrl_enable_state` functions, only initializes all the states as disabled. The functions are supposed to be used exactly the same way as before with the only difference that selected states need to be enabled at some time point during the initialization. Either by the Zephyr Power Management subsystem or the application itself. 

Regarding the comment
> But only one client can request it be enabled ("Any power state can be enabled only once."), which is enforced by an assert that the count never exceeds 1.

To ensure the client can operate undisturbed it may require (and request) some state to be disabled. The logic here is different than in case of a device driver that is used by multiple clients. In a typical scenario a low power state would turn off some high frequency clocks. An application (or driver) that requires the clock to be always enabled would need to disable the selected power state. Later when the application decides that all the devices dependent on the clock can be turned off the power state can be re-enabled. To sum up: a power state can be entered only if no one is blocking the entry.

**Comment by carlescufi** [comment]
@mnkp can you please rebase and add a bit of documentation regarding the change in behavior?
@wentongwu since @pabigot is currently neutral on this one, can you voice your opinion on whether this should go in?

**Review by pabigot — CHANGES_REQUESTED** [review]
I removed my NAK related to poor documentation of how this API is to be used correctly, since that's not the focus of the PR and the changes don't make it worse.

I note [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/21118#pullrequestreview-367175805):

> But it need current working platforms to do some extra initialization.

In short this just switches the default behavior, requiring changes to existing code that relies on the previous behavior, and to existing code that worked around previous behavior.

IMO this change would produce unnecessary churn given that we need to fix this right.

This is just an epsilon away from a NAK to me.  The only reason it isn't a NAK is because of an implied promise I wouldn't block it.  And on full consideration that's not a good enough excuse.  IMO this is not the right solution to the problem.

**Comment by mnkp** [comment]
> > But it need current working platforms to do some extra initialization.
> 
> In short this just switches the default behavior

That's exactly the problem, the default behavior was incorrect. We can't keep it. It definitely would be better to introduce the change along all the other rework that power management subsystem needs to undergo. Still, the change is a bug fix and without it the power management subsystem is broken. The system can enter sleep mode even if user didn't intended it. It's a rather serious deficiency. 

This part will not change in the future. We can't start with power states enabled before the power management system is all set up.

**Comment by wentongwu** [comment]
I agree remove SYS_PM_STATE_LOCK Kconfig option and enable the code under it by default. But IMHO to avoid unexpectedly power off system because of long timeout or sleep forever, we can separate system off state from deep sleep states as a individual and specific state so that it can be determined by user instead of residency policy, and this won't break the current working platform.

**Comment by pabigot** [comment]
@wentongwu Please either *approve* or *reject* this PR on the basis of what's in it, not on what could be in it.   We will be discussing this in dev-review tomorrow, and if the consensus is that the existing behavior is a bug that should be fixed *using this approach* I will remove my NAK and it will probably get merged as-is (after a rebase).

**Comment by pabigot** [comment]
I'd like to clarify my position on this PR in anticipation of dev-review discussion.

There is absolutely no question that a power management system that *by default* allows entry to deep sleep [*as currently defined*](https://docs.zephyrproject.org/latest/reference/power_management/index.html#terminology) based solely on the duration of expected sleep ("residency") is wrong.  This is not a behavior most developers would expect, and that should have been addressed during review.

However: This behavior was added in fall of 2018.  Existing code depends on this behavior, and existing code uses the API as implemented to avoid this behavior.

There has been churn in this API in the past: #22084 reversed a previous change in #20950 that removed deep-sleep states from the residency policy to solve this exact problem.  The reversal was required because existing code depends on a residency-based support for deep sleep, and because enabling deep-sleep states at all is opt-in.

The problematic state from #11908 may be called "system off" in Nordic terms but it is a true deep sleep state as currently defined by Zephyr: it loses execution context that must be recovered somehow.  It is simply a deep-sleep state that must not be entered using the residency policy because it cannot be exited based on the passage of time.  On other systems deep sleep states can safely be entered and exited based on residency.  This suggests the flaw is not in the states, but the set of states that are affected by the residency policy.

This PR proposes one way to change that: making the individual states opt-in.

My reject was influenced by the lack of non-comment reviews from people involved in the previous decisions including but not limited to @wentongwu @nashif @andrewboie @erwango @mnkp @vanti.  I'll replace my reject with approval if it seems we won't be re-re-reversing the default behavior in another six months.

**Review by wentongwu — CHANGES_REQUESTED** [review]
sorry for the confuse, but I don't think this PR is the best choice to the problem, let's review it in dev-review meeting.

**Comment by vanti** [comment]
Adding my two-cents. I tend to think of default behavior such as this as something that should only be changed as last resort - it can be error-prone for people who have been relying on the existing behavior. In this particular case, there seems to be a similar concern as I had regarding the deep sleep (shutdown) state on the TI CC13x2. The way I approached it was by defining our own residency policy (there were other reasons why I did so, but preventing the residency policy from entering our DEEP_SLEEP_1 state was one of them). If you look at `subsys/power/policy/policy_residency_cc13x2_cc26x2.c` , you would see I left out entries for deep sleep states in the pm_min_residency array and made sure sys_pm_policy_next_state() does not return the SYS_POWER_STATE_DEEP_SLEEP_1 state. Maybe this is an approach you can consider as an alternative to changing the defaults.

FWIW, I do like the idea of eliminating the SYS_PM_STATE_LOCK option. It seems to me this is functionality that is needed in most cases.

**Comment by mnkp** [comment]
I would like to clarify rationale for this PR in a bit more detail. There was a lot of talk regarding handling of deep sleep states since this is the issue reported by #11908. However, the problem is not limited to deep sleep states. Entering prematurely a sleep state can be just as deadly.

Looking at the Zephyr tree the SoC series/families that support System Power Management are: mec1501, nordic_nrf, silabs_exx32 and TI cc13x2_cc26x2. The Nordic supports deep sleep only. However, all the remaining SoC series/families have sleep states that turn off high frequency clocks. When System Power Manager schedules entry into a sleep state that turns off high frequency clock it ultimately shuts off most of the peripherals and (unless it was planned) renders the system unusable. Additionally, if we don't configure (or happen to configure) a wake up event the SoC will never leave the low power state it entered.

Automatic entry into a low power stated requires careful planning and coordination between System Power Management and Device Power Management subsystem (or an application if it's an application that wants to manage the system). Both need to be enabled and supported on a given SoC for automatic entry into low power state to make any sense.

At present only TI cc13x2_cc26x2 has support that looks complete enough that we may want to enable automatic entry into low power state during system initialization. @vanti would need to speak to that.

This PR doesn't aim to solve all the issues with the power management subsystem. And there are quite a few that are significant. We defined "Sleep State" as a state that doesn't lose execution context and "Deep Sleep State" as a state that does. However there are SoCs where CPU  may not lose execution context but peripherals are power gated and do lose configuration. We don't really know how to handle this scenario. So we had #22084, #20950 the churn that @pabigot was complaining about and are coming up with new terms like "system off" state that are not clearly defined. Also, the documentation talks about distributed and central method for device power management. While in fact we don't give user a clear choice to have the application or Zephyr manage the power. There are still a lot of issues that need to be resolved.

In any case, regardless if it's an application or Zephyr controlling power management the power states can only be enabled once the system and wake up events are properly configured. They simply cannot be enabled by default (or otherwise the policies cannot be enabled by default).

That said, I agree that it would make more sense to update Power Management Subsystem in one go and if there is no pressure on reducing the bug count we may well handle this issue later.

**Comment by galak** [comment]
Dev-review (May 28): Decided to defer for post 2.3

**Comment by vanti** [comment]
> At present only TI cc13x2_cc26x2 has support that looks complete enough that we may want to enable automatic entry into low power state during system initialization. @vanti would need to speak to that.

@mnkp From the cc13x2_cc26x2 standpoint I am not aware of any issue regarding automatic entry of power states in idle, which is why I am curious to learn more about the use-cases you have in mind. Is there a test/scenario you can share that would cause a premature entry to sleep? My understanding is that, short of the application forcing a (system) power state, the power state can only changed while the CPU is idle, which I'd expect to happen after initialization has completed and main() is entered. It should be fair for the policy function to run after that point in time and determine which state to transition to. And if for some reason we can end up in idle prior to main() before the devices are initialized, then it seems to me there may be a problem with the system's design.

> Looking at the Zephyr tree the SoC series/families that support System Power Management are: mec1501, nordic_nrf, silabs_exx32 and TI cc13x2_cc26x2. The Nordic supports deep sleep only. However, all the remaining SoC series/families have sleep states that turn off high frequency clocks. When System Power Manager schedules entry into a sleep state that turns off high frequency clock it ultimately shuts off most of the peripherals and (unless it was planned) renders the system unusable. 

To take an example, standby mode (SLEEP_2) on cc13x2_cc26x does indeed turn off most of the peripherals. Consequently it should only be entered when the peripherals are not needed. The way I conveyed this to power policy is by calling sys_pm_ctrl_enable/disable_state in the driver, to prevent SLEEP_2 from being entered while the peripheral is needed for an ongoing transfer for instance. So when a power state is enabled, it means it is ok for the policy to choose that state, and to "render the system unusable" since the CPU and the peripherals are not currently needed. Upon waking up the system would be restored to a state that is "usable" again.

> Additionally, if we don't configure (or happen to configure) a wake up event the SoC will never leave the low power state it entered.

Not sure why you wouldn't configure a wakeup source. Typically something like a timer or a gpio should be set up to wake up the system.

> Automatic entry into a low power stated requires careful planning and coordination between System Power Management and Device Power Management subsystem (or an application if it's an application that wants to manage the system). Both need to be enabled and supported on a given SoC for automatic entry into low power state to make any sense.

I agree with this. While separating system pm and device pm into two separate knobs seem intuitively appealing, in practice I found it is hard to live with one without the other. On cc13x2_cc26x2, I had to do some extra work to determine whether the user has only enabled system pm, and if so, I had to make sure to lean on the TI HAL to let me add a hook function that would reinitialize the peripheral when coming out of SLEEP_2, given device_set_power_state() would not be called. Doing system power management without regards to the devices isn't really realistic imho.

> However there are SoCs where CPU may not lose execution context but peripherals are power gated and do lose configuration. We don't really know how to handle this scenario. 

This is exactly the case we have for SLEEP_2 on cc13x2_cc26x2. We reinitialize the peripherals on exit of the sleep state.

> Also, the documentation talks about distributed and central method for device power management. While in fact we don't give user a clear choice to have the application or Zephyr manage the power. There are still a lot of issues that need to be resolved.

I agree this needs to be clarified. I was in same the boat when I first read the doc, and thought to myself: what is this distributed method it is talking about?  From what I understand central method is the only one that is supported today, and distributed method means "you are on your own" and "cook up your own power management in the application".

**Comment by mnkp** [comment]
> From the cc13x2_cc26x2 standpoint I am not aware of any issue regarding automatic entry of power states in idle, which is why I am curious to learn more about the use-cases you have in mind. Is there a test/scenario you can share that would cause a premature entry to sleep? My understanding is that, short of the application forcing a (system) power state, the power state can only changed while the CPU is idle, which I'd expect to happen after initialization has completed and main() is entered. It should be fair for the policy function to run after that point in time and determine which state to transition to. And if for some reason we can end up in idle prior to main() before the devices are initialized, then it seems to me there may be a problem with the system's design.

So indeed we can end up in idle prior to main(). The system power management `_sys_suspend()` function will be run the moment the kernel is initialized. Any code run in `POST_KERNEL` and later stage, including majority of the driver initialization code, can trigger entry into low power state. On TI cc13x2/cc26x2 this shouldn't be a major issue since you've implemented device power management code and wake up from idle. However, as I mentioned the TI cc13x2/cc26x2 is the only platform where power management implementation seems complete. On all other platforms any call to idle that is longer than the values defined by `CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_x` will put the SoC to sleep. We don't run into many issues with Zephyr's sample apps since the defaults for `CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_x` are unrealistically large 5, 10 and 30 s. Typically, depending on which parts of the clock tree are being turned off, entry and exit from a low power state takes between dozens of microseconds to a few milliseconds. So, in a real life application the values of `CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_x` will be in range of ms. It's likely to have small delays in the range of milliseconds in the driver code which would be enough to trigger entry into low power state.

Even on TI cc13x2/cc26x2 platform with its pretty complete power management support things will break quickly if the device power management code is not enabled. Enabling System Power Management but not Device Power Management is one way how user can configure Zephyr to let the application control entry into power states.

Things will also break if `CONFIG_SYS_POWER_DEEP_SLEEP_STATES` is enabled and user's application runs some initialization code at `POST_KERNEL` stage, prior to driver initialization, that calls `k_sleep` with value large enough to trigger entry into low power mode. That's not an unusual scenario. In a low power design it is quite common to power gate external components on the PCB, e.g. external flash, sensors that may be connected via I2C bus, etc. Application that wants to initialize those components during system boot would need to enable first respective power supplies. This would happen after the kernel is initialized but before the driver initialization. A call to  `k_sleep` would be there to wait until power supplies stabilize.

> > Additionally, if we don't configure (or happen to configure) a wake up event the SoC will never leave the low power state it entered.
> 
> Not sure why you wouldn't configure a wakeup source. Typically something like a timer or a gpio should be set up to wake up the system.

We should of course configure a wake up source but this is not done automatically on all the platforms. E.g. SiLabs is using standard SysTick running on high frequency clock which is turned off the moment it enters a low power state. On this platform the Device Power Management is not implemented. It's a responsibility of the application to manage entry into low power states and set up a wake up source.

**Comment by vanti** [comment]
> So indeed we can end up in idle prior to main(). The system power management `_sys_suspend()` function will be run the moment the kernel is initialized. Any code run in `POST_KERNEL` and later stage, including majority of the driver initialization code, can trigger entry into low power state.

@mnkp Thanks for the detailed explanation. The ability to switch to the idle thread before entering main() is something I am not used to, coming from my prior experience on TI-RTOS, where interrupts and threads aren't enabled til exiting main(). I'd have thought calling k_sleep()/k_sem_take() etc. would at least be frowned upon during driver initialization. Meanwhile, I personally haven't seen _sys_suspend() being called prior to main() while working on the cc1352r1_launchxl, given the existing drivers do not call functions that block the main thread during initialization.

In any event, it sounds like the issue you are facing is that the code responsible for setting up the devices that can act as wake-up sources doesn't get run prior to entering idle, potentially leading to a permanent sleep. Thinking out loud, instead of changing the defaults, there seem to be a few possibilities:

1. Change the drivers to not make calls that would block the main thread

2. Change the drivers of the devices that could act as wake-up sources to be initialized prior to POST_KERNEL initialization. Or maybe they can be prioritized to be initialized earlier in POST_KERNEL init, before the drivers that make blocking calls

3. Plug a PRE_KERNEL init function to call sys_pm_ctrl_disable_state() on the states that should not be entered during initialization, and re-enable them later after the drivers are initialized and when it is safe. This allows "lighter" sleep states that aren't relying on power-gating for example to stay enabled throughout, thereby saving power when k_sleep() or some other blocking call is made during driver initialization, which I think is the goal given it was preferred to k_busy_wait().

4. Add a system pm policy function that distinguishes initialization stage vs. when it is safe to sleep more deeply, and enter appropriate sleep states accordingly.

Would any of these (or a combination) work? Again, just throwing out some ideas.

**Pull request closed without merging**